conformance/test-server/src/main.rs

```diff
@@ -0,0 +1,298 @@
+use std::net::SocketAddr;
+
+use anyhow::Result;
+use clap::Parser;
+use futures::future;
+use hickory_proto::op::Message;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, UdpSocket},
+    task::JoinHandle,
+};
+
+mod handlers;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let args = Args::parse();
+    let transport = args.transport;
+    let handler = match &args.handler[..] {
+        "base" => handlers::base_handler,
+        _ => {
+            return Err(anyhow::Error::msg("unknown handler"));
+        }
+    };
+
+    let mut handles = vec![];
+    if transport == "tcp" || transport == "both" {
+        let (_, tcp_handle) = tcp_server(([0, 0, 0, 0], args.port).into(), handler).await?;
+        handles.push(tcp_handle);
+    }
+    if transport == "udp" || transport == "both" {
+        let (_, udp_handle) = udp_server(([0, 0, 0, 0], args.port).into(), handler).await?;
+        handles.push(udp_handle);
+    }
+    let _ = future::join_all(handles).await;
+
+    Ok(())
+}
+
+#[derive(Parser)]
+#[clap(name = "Message responder test DNS server")]
+struct Args {
+    #[clap(default_value = "base", long = "handler")]
+    handler: String,
+    #[clap(default_value = "53", long = "port")]
+    port: u16,
+    #[clap(default_value = "both", long = "transport")]
+    transport: String,
+}
+
+async fn udp_server(
+    bind_addr: SocketAddr,
+    handler: HandlerFn,
+) -> Result<(SocketAddr, JoinHandle<()>)> {
+    let udp = UdpSocket::bind(bind_addr).await?;
+    let addr = udp.local_addr()?;
+    let handle = tokio::task::spawn(async move {
+        loop {
+            let mut read_buf = [0u8; 4096];
+            let to = match udp.recv_from(&mut read_buf).await {
+                Ok((_, to)) => to,
```

Comment:
I think we need the `usize` value here to find out the length of the message. As it currently is, we'll pass all 4096 bytes of `read_buf` to `Message::from_vec()` which I guess the DNS format can deal with but might have surprising results.